### PR TITLE
Make some realtime API actions available to plugins

### PR DESCRIPTION
**Pull request by thomasarbona**
:warning: depends on https://github.com/kuzzleio/kuzzle/pull/1358

## What does this PR do ?

Add availability of some realtime methods in plugins (all instead of `subscribe` and `unsubscribe`).

### How should this be manually tested?

- create a plugin
- use some method of the realtime controller

### CODE REVIEW

File: lib/api/core/plugins/pluginContext.js [hunk 1]
```diff
@@ -138,6 +138,10 @@ class PluginContext {
         get: () => curryTrigger(kuzzle, pluginName)
       });
 
+      const throwNotAvailable = (name) => () => {
```
Comment (kuzzle) [inline lib/api/core/plugins/pluginContext.js]: MINOR Remove parentheses around the parameter of this arrow function. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS3524)

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: lib/api/core/plugins/pluginContext.js [hunk 2]
```diff
@@ -238,8 +241,8 @@ function execute (kuzzle, request, callback) {
     return deferred;
   }
 
-  if (request.input.controller === 'realtime') {
-    error = new PluginImplementationError('Realtime controller is not available in plugins. You should use plugin hooks instead');
+  if (request.input.controller === 'realtime' && ['subscribe', 'unsubscribe'].includes(request.input.action)) {
```
Comment (Aschen) [inline lib/api/core/plugins/pluginContext.js]: `Array.includes` won't work in Node.js 6

File: lib/api/core/sdk/funnelProtocol.js [hunk 3]
```diff
@@ -59,8 +59,8 @@ class FunnelProtocol extends KuzzleEventEmitter {
       user: this.user
     });
 
-    if (kuzzleRequest.input.controller === 'realtime') {
-      return Bluebird.reject(new PluginImplementationError('Realtime controller is not available in plugins. You should use plugin hooks instead'));
+    if (kuzzleRequest.input.controller === 'realtime' && ['subscribe', 'unsubscribe'].includes(kuzzleRequest.input.action)) {
```
Comment (Aschen) [inline lib/api/core/sdk/funnelProtocol.js]: ditto

File: lib/api/core/plugins/pluginContext.js [hunk 4]
```diff
@@ -146,10 +150,9 @@ class PluginContext {
         document: sdk.document,
         index: sdk.index,
         ms: sdk.ms,
-        realtime: new Proxy({}, {
-          get () {
-            throw new PluginImplementationError('Realtime controller is not available in plugins. You should use plugin hooks instead');
-          }
+        realtime: Object.assign(sdk.realtime, {
+          subscribe: throwNotAvailable('realtime.subscribe'),
```
Comment (Aschen) [inline lib/api/core/plugins/pluginContext.js]: Nit: we use `controller:action` notation in the rest of the documentation/kuzzle
```suggestion
          subscribe: throwNotAvailable('realtime:subscribe'),
```

File: lib/api/core/plugins/pluginContext.js [hunk 5]
```diff
@@ -146,10 +150,9 @@ class PluginContext {
...
+          unsubscribe: throwNotAvailable('realtime.unsubscribe')
```
Comment (Aschen) [inline lib/api/core/plugins/pluginContext.js]:
```suggestion
          unsubscribe: throwNotAvailable('realtime:unsubscribe')
```

File: test/api/core/pluginContext/pluginContext.test.js [hunk 6]
```diff
@@ -478,9 +497,9 @@ describe('Plugin Context', () => {
       it('should reject if trying to call the realtime controller', () => {
         return should(context.accessors.execute(new Request({
           controller: 'realtime',
-          action: 'publish'
+          action: 'subscribe'
         })))
-          .be.rejectedWith(/Realtime controller is not available in plugins\. You should use plugin hooks instead/);
+          .be.rejectedWith(/realtime.subscribe method is not available in plugins\. You should use plugin hooks instead/);
```
Comment (Aschen) [inline test/api/core/pluginContext/pluginContext.test.js]:
```suggestion
          .be.rejectedWith(/realtime:subscribe method is not available in plugins\. You should use plugin hooks instead/);
```

File: test/api/core/sdk/funnelProtocol.test.js [hunk 7]
```diff
@@ -82,9 +82,9 @@ describe('Test: sdk/funnelProtocol', () => {
     it('should reject if trying to call the realtime controller', () => {
       return should(funnelProtocol.query({
         controller: 'realtime',
-        action: 'publish'
+        action: 'subscribe'
       }))
-        .be.rejectedWith(/Realtime controller is not available in plugins\. You should use plugin hooks instead/);
+        .be.rejectedWith(/realtime.subscribe method is not available in plugins\. You should use plugin hooks instead/);
```
Comment (Aschen) [inline test/api/core/sdk/funnelProtocol.test.js]:
```suggestion
        .be.rejectedWith(/realtime:subscribe method is not available in plugins\. You should use plugin hooks instead/);
```

File: lib/api/core/plugins/pluginContext.js (hunk 2 above)
Comment (thomasarbona) [reply lib/api/core/plugins/pluginContext.js]: https://node.green/#ES2016-features-Array-prototype-includes-Array-prototype-includes

Comment (berthieresteban) [inline lib/api/core/plugins/pluginContext.js]: 80 cols

File: lib/api/core/plugins/pluginContext.js [hunk 8]
```diff
@@ -138,6 +149,15 @@ class PluginContext {
         get: () => curryTrigger(kuzzle, pluginName)
       });
 
+      const throwNotAvailable = (name) => () => {
```
Comment (kuzzle) [inline lib/api/core/plugins/pluginContext.js]: MINOR Remove parentheses around the parameter of this arrow function. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS3524)

**Comment by kuzzle** [comment]
SonarQube analysis reported 2 issues
* ![MINOR][MINOR] 1 minor
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [pluginsManager.js#L1011](https://github.com/kuzzleio/kuzzle/blob/61d9a9a4fd3d2938b0f9a9eb42a3224f72add853/lib/api/core/plugins/pluginsManager.js#L1011): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [pluginsManager.js#L1011](https://github.com/kuzzleio/kuzzle/blob/802cad3d19c1247ab8df8a25f06b4acf1b97cff1/lib/api/core/plugins/pluginsManager.js#L1011): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [pluginsManager.js#L1011](https://github.com/kuzzleio/kuzzle/blob/725dfd76686c2a9816676e51bd4e76fe3ac71fd4/lib/api/core/plugins/pluginsManager.js#L1011): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [pluginsManager.js#L1011](https://github.com/kuzzleio/kuzzle/blob/0d8802ae14d50010b0aacacc1a5be243a8302615/lib/api/core/plugins/pluginsManager.js#L1011): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [pluginsManager.js#L1011](https://github.com/kuzzleio/kuzzle/blob/0b345e100a5b566ceaf3599565bd3c3f11287024/lib/api/core/plugins/pluginsManager.js#L1011): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [pluginsManager.js#L1011](https://github.com/kuzzleio/kuzzle/blob/5395e79f94240f90d38c85ab5dd62c1f5bec8b23/lib/api/core/plugins/pluginsManager.js#L1011): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [pluginsManager.js#L1011](https://github.com/kuzzleio/kuzzle/blob/7ae6df897aa6b665755c7386c224585f83a86079/lib/api/core/plugins/pluginsManager.js#L1011): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [pluginsManager.js#L1011](https://github.com/kuzzleio/kuzzle/blob/0ba482af66647af005950a8c8f3cacb7a5170bbe/lib/api/core/plugins/pluginsManager.js#L1011): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [pluginsManager.js#L1011](https://github.com/kuzzleio/kuzzle/blob/6379b14b81285d4b60fd3ca230cfda0309cd8a25/lib/api/core/plugins/pluginsManager.js#L1011): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [pluginsManager.js#L1003](https://github.com/kuzzleio/kuzzle/blob/ccdc7822c65bdb18458700449674c647997b9578/lib/api/core/plugins/pluginsManager.js#L1003): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

File: lib/api/core/plugins/pluginContext.js [hunk 9]
```diff
@@ -148,6 +148,14 @@ class PluginContext {
         get: () => curryTrigger(kuzzle, pluginName)
       });
 
+      const throwNotAvailable = name => () => {
```
Comment (kuzzle) [inline lib/api/core/plugins/pluginContext.js]: MAJOR Remove this useless assignment to local variable "throwNotAvailable" [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1854)
MINOR Remove the declaration of the unused 'throwNotAvailable' variable. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AUnusedVariable)

**Comment by kuzzle** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**